x-pack/plugin/slm/src/internalClusterTest/java/org/elasticsearch/xpack/slm/SLMStatDisruptionIT.java

```diff
@@ -0,0 +1,638 @@
+            assertMetadata(policyName, 2, 0, 0, Set.of());
+
+        }, 1, TimeUnit.MINUTES);
+    }
+
+    /**
+     * Test that after successful snapshot preRegisteredRuns status is 0.
+     */
+    public void testSuccessSnapshot() throws Exception {
+        final String idxName = "test-idx";
+        final String repoName = "test-repo";
+        final String policyName = "test-policy";
+
+        internalCluster().startMasterOnlyNodes(1);
+        final String masterNode = internalCluster().getMasterName();
+        final String dataNode = internalCluster().startDataOnlyNode();
+        ensureStableCluster(2);
+
+        createRandomIndex(idxName, dataNode);
+        createRepository(repoName, "mock");
+        createSnapshotPolicy(policyName, "snap", NEVER_EXECUTE_CRON_SCHEDULE, repoName, idxName);
+
+        ensureGreen();
+
+        String snapshotName = executePolicy(masterNode, policyName);
+        logger.info("Created snapshot: " + snapshotName);
+
+        waitForSnapshot(repoName, snapshotName);
+
+        assertBusy(() -> {
+            assertSnapshotSuccess(repoName, snapshotName);
+            assertMetadata(policyName, 1, 0, 0, Set.of());
+        }, 1, TimeUnit.MINUTES);
+    }
+
+    /**
+     * Test that after a failure then a success, preRegisteredRuns from failure is added to invocationsSinceLastSuccess.
+     */
+    public void testFailSnapshotFailStatsThenSuccessRecoverStats() throws Exception {
+        final String idxName = "test-idx";
+        final String repoName = "test-repo";
+        final String policyName = "test-policy";
+
+        internalCluster().startMasterOnlyNodes(1);
+        final String masterNode = internalCluster().getMasterName();
+        final String dataNode = internalCluster().startDataOnlyNode();
+        ensureStableCluster(2);
+
+        NetworkDisruption networkDisruption = isolateMasterDisruption(NetworkDisruption.DISCONNECT);
+        internalCluster().setDisruptionScheme(networkDisruption);
+
+        // Listener that stops disrupting network only after snapshot completion
+        CountDownLatch latch = new CountDownLatch(1);
+        internalCluster().clusterService(masterNode).addListener(new WaitForSnapshotListener(repoName, networkDisruption, latch));
+
+        createRandomIndex(idxName, dataNode);
+        createRepository(repoName, "mock");
+        createSnapshotPolicy(policyName, "snap", NEVER_EXECUTE_CRON_SCHEDULE, repoName, idxName);
+
+        ensureGreen();
+
+        networkDisruption.startDisrupting();
+        String snapshotName = executePolicy(masterNode, policyName);
+        logger.info("Created snapshot: " + snapshotName);
+
+        // wait for snapshot to complete and network disruption to stop
+        assertTrue(latch.await(1, TimeUnit.MINUTES));
+
+        // restart master so failure stat is lost
+        // TODO this relies on a race condition.
+        // The node restart must happen before stats are stored in cluster state, but this is not guaranteed.
```

Comment:
Are these TODOs still relevant? Is this going to cause failures down the road?

Reply:
The race condition is still present ... but I've never actually seen it not fail to lose the stats. I think it's probably fine to remove the TODO, but leave a big comment saying: "if this test fails it might be because of this race condition"

Reply:
Do you mean that you've never seen it lose the stats? Or do you mean when it occurs it always loses the stats? If it hasn't failed then removing the TODOs (I think there are a couple of them) and replacing with the other comment sounds good to me

Reply:
Yeah, the intention of that line it to force it to lose stats. So far, it has always succeeded in losing the stats. Cool, then I'll remove the todo.